Packs/Expanse/Integrations/Expanse/Expanse.py

```diff
@@ -0,0 +1,541 @@
+import demistomock as demisto
+from CommonServerPython import *
+from CommonServerUserPython import *
+
+''' IMPORTS '''
+
+import json
+import requests
+from datetime import datetime, timedelta
+
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()
+
+''' GLOBALS/PARAMS '''
+
+
+PARAMS = demisto.params()
+API_KEY = PARAMS.get('api_key')
+PAGE_LIMIT = PARAMS.get('page_limit')
+FIRST_RUN = PARAMS.get('first_run')
+SERVER = 'https://expander.expanse.co'
+INSECURE = PARAMS.get('insecure')
+PROXY = demisto.params().get('proxy')
+BASE_URL = SERVER
+EXPOSURE_EVENT_TYPES = "ON_PREM_EXPOSURE_APPEARANCE,ON_PREM_EXPOSURE_REAPPEARANCE"
+API_ENDPOINTS = {
+    "exposures/ip-ports": {
+        "version": 2
+    },
+    "ip-range": {
+        "version": 2
+    },
+    "assets/domains": {
+        "version": 2
+    },
+    "IdToken": {
+        "version": 1
+    },
+    "events": {
+        "version": 1
+    }
+}
+
+EXPOSURE_SEVERITY_MAPPING = {
+    "NONE": 0,
+    "UNKNOWN": 0,
+    "CRITICAL": 3,
+    "ROUTINE": 1,
+    "WARNING": 2,
+    "UNCATEGORIZED": 1
+}
+
+''' HELPER FUNCTIONS '''
+
+
+def make_headers(endpoint, token):
+    ''' provides proper headers for differing authentication methods to API '''
+    headers = {
+        'Content-Type': 'application/json',
+        'Accept': 'application/json'
+    }
+    if API_ENDPOINTS[endpoint]['version'] == 1:
+        headers['Authorization'] = 'Bearer ' + token
+
+    elif API_ENDPOINTS[endpoint]['version'] == 2:
+        headers['Authorization'] = 'JWT ' + token
+
+    return headers
+
+
+def make_url(endpoint):
+    ''' build URL based on endpoint'''
+    url = "{BASE_URL}/api/v{version}/{endpoint}".format(
+        BASE_URL=BASE_URL,
+        version=API_ENDPOINTS[endpoint]['version'],
+        endpoint=endpoint
+    )
+    return url
+
+
+def get_page_token(url):
+    o = {'pagetoken': False}
+    for i in url.split("&"):
+        r = i.split("=")
+        o[r[0]] = r[1]
+    return o['pageToken']
+
+
+def do_auth():
+    ''' perform authentication using API_KEY,
+    stores token and stored timestamp in integration context,
+    retrieves new token when expired '''
+    auth = demisto.getIntegrationContext()
+    LOG("Integration context")
+    LOG(auth)
+    now_epoch = int(datetime.today().strftime('%s'))
+
+    try:
+        "token" in auth
+        "stored" in auth
+        int(auth['stored']) + (60 * 60 * 4) > int(now_epoch)
```

Comment:
what do those checks do? They will not get us to the except section

Reply:
lines 99 and 100 are for the first time the app runs, line 101 checks that the JWT token is less than 4 hours old, and if not will force fetching a new JWT token ( these expire every 6 hours)